plugins/mana/plugin.go

```diff
@@ -371,8 +413,299 @@ func GetPendingMana(value float64, n time.Duration) float64 {
 	return value * (1 - math.Pow(math.E, -mana.Decay*(n.Seconds())))
 }
 
+// GetLoggedEvents gets the events logs for the node IDs and time frame specified. If none is specified, it returns the logs for all nodes.
+func GetLoggedEvents(IDs []identity.ID, startTime time.Time, endTime time.Time) (map[identity.ID]*EventsLogs, error) {
+	logs := make(map[identity.ID]*EventsLogs)
+	lookup := make(map[identity.ID]bool)
+	getAll := true
+
+	if len(IDs) > 0 {
+		getAll = false
+		for _, nodeID := range IDs {
+			lookup[nodeID] = true
+		}
+	}
+
+	var err error
+	consensusEventsLogStorage.ForEach(func(key []byte, cachedObject objectstorage.CachedObject) bool {
+		cachedPe := &mana.CachedPersistableEvent{CachedObject: cachedObject}
+		defer cachedPe.Release()
+		pbm := cachedPe.Unwrap()
+
+		if !getAll {
+			if !lookup[pbm.NodeID] {
+				return true
+			}
+		}
+
+		if _, found := logs[pbm.NodeID]; !found {
+			logs[pbm.NodeID] = &EventsLogs{}
+		}
+
+		var ev mana.Event
+		ev, err = mana.FromPersistableEvent(pbm)
+		if err != nil {
+			return false
+		}
+
+		pe := ev.ToPersistable()
+		if pe.Time.Before(startTime) || pe.Time.After(endTime) {
+			return true
+		}
+		switch ev.Type() {
+		case mana.EventTypePledge:
+			logs[pbm.NodeID].Pledge = append(logs[pbm.NodeID].Pledge, ev.(*mana.PledgedEvent))
+		case mana.EventTypeRevoke:
+			logs[pbm.NodeID].Revoke = append(logs[pbm.NodeID].Revoke, ev.(*mana.RevokedEvent))
+		default:
+			err = mana.ErrUnknownManaEvent
+			return false
+		}
+		return true
+	})
+
+	for ID := range logs {
+		sort.Slice(logs[ID].Pledge, func(i, j int) bool {
+			return logs[ID].Pledge[i].Time.Before(logs[ID].Pledge[j].Time)
+		})
+		sort.Slice(logs[ID].Revoke, func(i, j int) bool {
+			return logs[ID].Revoke[i].Time.Before(logs[ID].Revoke[j].Time)
+		})
+	}
+
+	return logs, err
+}
+
+// GetPastConsensusManaVectorMetadata gets the past consensus mana vector metadata.
+func GetPastConsensusManaVectorMetadata() mana.ConsensusBasePastManaVectorMetadata {
+	cachedObj := consensusBaseManaPastVectorMetadataStorage.Get([]byte(mana.ConsensusBaseManaPastVectorMetadataStorageKey))
+	cachedMetadata := &mana.CachedConsensusBasePastManaVectorMetadata{CachedObject: cachedObj}
+	defer cachedMetadata.Release()
+	metadata := cachedMetadata.Unwrap()
+	return *metadata
```

Comment:
when we don't have a saved metadata, dereferencing the nil pointer causes a panic. Check for `nil` in the return of `Unwrap()`.